keeper/abci.go

```diff
@@ -22,12 +22,172 @@ package keeper
+// positions proportionally to their share holdings.
+func (k *Keeper) updateVaultsV2Accounting(ctx context.Context) error {
+	navInfo, err := k.GetVaultsV2NAVInfo(ctx)
+	if err != nil {
+		return err
+	}
+
+	if navInfo.CurrentNav.IsNil() {
+		return nil
+	}
+
+	state, err := k.GetVaultsV2VaultState(ctx)
+	if err != nil {
+		return err
+	}
+
+	totalShares, err := k.GetVaultsV2TotalShares(ctx)
+	if err != nil {
+		return err
+	}
+
+	// When there are no active shares, reset accrued yield information and
+	// simply mirror the NAV in the aggregate vault state.
+	if totalShares.IsZero() {
+		if err := k.IterateVaultsV2UserPositions(ctx, func(address types.AccAddress, position vaultsv2.UserPosition) (bool, error) {
+			if position.AccruedYield.IsZero() {
+				return false, nil
+			}
+
+			position.AccruedYield = sdkmath.ZeroInt()
+			if err := k.SetVaultsV2UserPosition(ctx, address, position); err != nil {
+				return true, err
+			}
+
+			return false, nil
+		}); err != nil {
+			return err
+		}
+
+		state.TotalNav = navInfo.CurrentNav
+
+		if !state.TotalDeposits.IsNil() {
+			if state.TotalAccruedYield, err = navInfo.CurrentNav.SafeSub(state.TotalDeposits); err != nil {
+				return err
+			}
+		} else {
+			state.TotalAccruedYield = sdkmath.ZeroInt()
+		}
+
+		if !navInfo.LastUpdate.IsZero() {
+			state.LastNavUpdate = navInfo.LastUpdate
+		}
+
+		return k.SetVaultsV2VaultState(ctx, state)
+	}
+
+	if state.TotalNav.Equal(navInfo.CurrentNav) && !navInfo.LastUpdate.After(state.LastNavUpdate) {
+		return nil
+	}
+
+	navBig := navInfo.CurrentNav.BigInt()
+	totalSharesBig := totalShares.BigInt()
+
+	residual := new(big.Int)
+	aggregatedYield := sdkmath.ZeroInt()
+	aggregatedDeposits := sdkmath.ZeroInt()
+
+	err = k.IterateVaultsV2UserPositions(ctx, func(address types.AccAddress, position vaultsv2.UserPosition) (bool, error) {
+		shares, err := k.GetVaultsV2UserShares(ctx, address)
+		if err != nil {
+			return true, err
+		}
+
+		if !shares.IsPositive() {
+			pending := position.AmountPendingWithdrawal
+
+			if !position.AccruedYield.IsZero() || position.DepositAmount.IsNil() || !position.DepositAmount.Equal(pending) {
+				position.AccruedYield = sdkmath.ZeroInt()
+				position.DepositAmount = pending
+				if err := k.SetVaultsV2UserPosition(ctx, address, position); err != nil {
+					return true, err
+				}
+			}
+
+			aggregatedDeposits, err = aggregatedDeposits.SafeAdd(pending)
+			if err != nil {
+				return true, err
+			}
+
+			return false, nil
+		}
+
+		numerator := new(big.Int).Mul(navBig, shares.BigInt())
+		quotient := new(big.Int)
+		remainder := new(big.Int)
+		quotient.QuoRem(numerator, totalSharesBig, remainder)
+
+		residual.Add(residual, remainder)
+		if residual.Cmp(totalSharesBig) >= 0 {
+			extra := new(big.Int).Div(new(big.Int).Set(residual), totalSharesBig)
+			if extra.Sign() > 0 {
+				quotient.Add(quotient, extra)
+				residual.Sub(residual, new(big.Int).Mul(totalSharesBig, extra))
+			}
+		}
```

Comment:
Do we need to check if `residual > 0` after the last iteration? Could we be losing dust if it is?